apps/nextjs/src/stores/resourcesStore/utils/callWithHandshakeRetry.ts

```diff
@@ -0,0 +1,163 @@
+import { aiLogger } from "@oakai/logger";
+
+const log = aiLogger("additional-materials");
+const wait = (ms: number) => new Promise((r) => setTimeout(r, ms));
+
+// Type guard for tRPC errors
+function isTRPCError(error: unknown): error is { data?: { code?: string } } {
+  return typeof error === "object" && error !== null && "data" in error;
+}
+
+// Type guard for HTTP errors
+function isHTTPError(error: unknown): error is { status?: number } {
+  return typeof error === "object" && error !== null && "status" in error;
+}
+
+// Type guard for TRPCClientError with HTML response (session not ready)
+function isTRPCHTMLError(error: unknown): error is { message: string } {
+  log.info("🔍 Checking if error is HTML response error", {
+    errorType: typeof error,
+    hasMessage:
+      typeof error === "object" && error !== null && "message" in error,
+    errorString: String(error).substring(0, 200), // First 200 chars for logging
+  });
+
+  const isHTMLError =
+    typeof error === "object" &&
+    error !== null &&
+    "message" in error &&
+    typeof (error as { message: unknown }).message === "string" &&
+    (error as { message: string }).message.includes(
+      "Unexpected token '<', \"<!DOCTYPE",
+    );
+
+  if (isHTMLError) {
+    log.info("✅ Detected HTML response error", {
+      message: (error as { message: string }).message,
+    });
+  } else {
+    log.info("❌ Not an HTML response error");
+  }
+
+  return isHTMLError;
+}
+
+// Helper to determine error types and retry strategy
+function analyzeError(error: unknown) {
+  const unauthorized =
+    (isTRPCError(error) && error.data?.code === "UNAUTHORIZED") ||
+    (isHTTPError(error) && error.status === 401);
+
+  const sessionNotReady = isTRPCHTMLError(error);
+
+  return { unauthorized, sessionNotReady };
+}
+
+// Helper to calculate retry delay based on error type and attempt
+function calculateDelay(
+  errorAnalysis: { unauthorized: boolean; sessionNotReady: boolean },
+  attempt: number,
+  baseDelayMs: number,
+): number {
+  const { unauthorized, sessionNotReady } = errorAnalysis;
+
+  if (unauthorized) {
+    return baseDelayMs;
+  } else if (sessionNotReady) {
+    return attempt * 500; // 500ms, 1000ms, 1500ms for session propagation
+  } else {
+    return baseDelayMs * Math.pow(2, attempt - 1);
+  }
+}
```

Comment:
We could avoid some of the logic in this file with a package like [p-retry](https://github.com/sindresorhus/p-retry)